File: Python/4-Stack/Problems/4-LargestRectangle-in-Histogram/LargestRectangle-in-HistogramPro.py
'''
Advanced Solutions for Largest Rectangle in Histogram - Real-World Applications

Problem Statement:
Given an array representing the heights of histogram bars where the width of each bar is 1,
return the area of the largest rectangle in the histogram.
'''

import heapq

# 1. Real-time Stock Market Analysis (Segment Tree Approach - O(N log N))
class SegmentTree:
    """
    Implements a Segment Tree to efficiently find the minimum height bar within a given range.
    This is useful in real-time stock market analysis to identify price ranges with minimum volatility,
    which can indicate support or resistance levels.  The Segment Tree allows for fast range queries.
    """
    def __init__(self, heights):
        """
        Initializes the SegmentTree object.

        Args:
            heights: A list of stock prices (or other relevant financial data).
        """
        self.n = len(heights)
        self.heights = heights
        self.tree = [0] * (4 * self.n)  # Allocate memory for the segment tree.
        self._build(0, 0, self.n - 1)  # Build the segment tree.

    def _build(self, node, start, end):
        """
        Recursively builds the segment tree.

        Args:
            node: The index of the current node in the tree.
            start: The starting index of the current range.
            end: The ending index of the current range.
        """
        if start == end:
            self.tree[node] = start  # Leaf node: store the index of the bar.
        else:
            mid = (start + end) // 2
            left_child = 2 * node + 1
            right_child = 2 * node + 2
            self._build(left_child, start, mid)  # Build the left subtree.
            self._build(right_child, mid + 1, end)  # Build the right subtree.
            # Internal node: store the index of the minimum height bar in the range.
            self.tree[node] = min(self.tree[left_child], self.tree[right_child], key=lambda i: self.heights[i])

    def query(self, node, start, end, l, r):
        """
        Performs a range query on the segment tree to find the minimum height bar index.

        Args:
            node: The index of the current node in the tree.
            start: The starting index of the current range.
            end: The ending index of the current range.
            l: The left boundary of the query range.
            r: The right boundary of the query range.

        Returns:
            The index of the minimum height bar in the query range, or -1 if the range is invalid.
        """
        if l > end or r < start:
            return -1  # Out of range.
        if l <= start and end <= r:
            return self.tree[node]  # Current range is within the query range.
        mid = (start + end) // 2
        left_index = self.query(2 * node + 1, start, mid, l, r)  # Query the left subtree.
        right_index = self.query(2 * node + 2, mid + 1, end, l, r)  # Query the right subtree.
        if left_index == -1:
            return right_index  # Only right subtree has valid results.
        if right_index == -1:
            return left_index  # Only left subtree has valid results.
        # Return the index of the smaller height.
        return left_index if self.heights[left_index] < self.heights[right_index] else right_index

def largestRectangleArea_segment_tree(heights):
    """
    Calculates the largest rectangle area in a histogram using a Segment Tree.
    This approach is efficient for scenarios where the histogram data is frequently updated,
    as the Segment Tree allows for efficient range minimum queries.

    Args:
        heights: A list of integers representing the heights of the histogram bars.

    Returns:
        The area of the largest rectangle in the histogram.
    """
    if not heights:
        return 0
    tree = SegmentTree(heights)  # Build the Segment Tree.

    def compute_area(left, right):
        """
        Recursively computes the largest rectangle area within the given range using the Segment Tree.

        Args:
            left: The left index of the range.
            right: The right index of the range.

        Returns:
            The area of the largest rectangle within the range.
        """
        if left > right:
            return 0
        min_index = tree.query(0, 0, len(heights) - 1, left, right)  # Use the Segment Tree to find the minimum.
        return max(
            heights[min_index] * (right - left + 1),
            compute_area(left, min_index - 1),
            compute_area(min_index + 1, right)
        )

    return compute_area(0, len(heights) - 1)

# 2. E-commerce Sales Distribution (Heap-Based - O(N log N))
def largestRectangleArea_heap(heights):
    """
    Calculates the largest rectangle area in a histogram using a heap.
    In the context of e-commerce, this can be used to analyze sales distribution,
    where bar heights represent sales volume for different product categories,
    and we want to find the largest rectangular area representing a significant sales segment.
    The heap is used to efficiently find the minimum height.

    Args:
        heights: A list of integers representing sales volumes for different product categories.

    Returns:
        The area of the largest rectangle in the histogram.
    """
    max_area = 0
    heap = []  # A min-heap to store (height, index) tuples.
    for i, h in enumerate(heights):
        heapq.heappush(heap, (h, i))  # Push the height and index onto the heap.
        while heap and heap[0][0] > h:
            # While the heap is not empty and the minimum height in the heap is greater than the current height...
            height, index = heapq.heappop(heap)  # ...pop the (height, index) tuple with the smallest height.
            width = i - index
            max_area = max(max_area, height * width)
    return max_area

# 3. Land Usage Analysis (Simple Sorting - O(N log N))
def largestRectangleArea_sorted(heights):
    """
    Calculates the largest rectangle area in a histogram using sorting.
    For land usage analysis, the heights might represent the area of land used for different purposes,
    and finding the largest rectangle can help identify the most significant land usage pattern.
    Sorting the heights allows us to process them in increasing order.

    Args:
        heights: A list of integers representing land areas for different purposes.

    Returns:
        The area of the largest rectangle in the histogram.
    """
    indexed_heights = sorted((h, i) for i, h in enumerate(heights))  # Sort (height, index) tuples by height.
    max_area = 0
    for h, i in indexed_heights:
        left = i
        right = i
        while left > 0 and heights[left - 1] >= h:
            left -= 1
        while right < len(heights) - 1 and heights[right + 1] >= h:
            right += 1
        max_area = max(max_area, h * (right - left + 1))
    return max_area

# 4. City Infrastructure Planning (Greedy with Two Pointers - O(N))
def largestRectangleArea_two_pointers(heights):
    """
    Calculates the largest rectangle area using a greedy approach with two pointers.
    In city infrastructure planning, the heights could represent building heights,
    and the largest rectangle can help determine optimal placement of structures to maximize space utilization
    or view.  However, this implementation has a flaw:  It doesn't work correctly for all cases.
    It's included to illustrate an incorrect approach.
    """
    max_area = 0
    left, right = 0, len(heights) - 1
    while left < right:
        height = min(heights[left], heights[right])
        max_area = max(max_area, height * (right - left + 1))
        if heights[left] < heights[right]:
            left += 1
        else:
            right -= 1
    return max_area # This method is incorrect.

# 5. Climate Data Analysis (Prefix Sum - O(N))
def largestRectangleArea_prefix_sum(heights):
    """
    Calculates the largest rectangle area using prefix sums.
    For climate data analysis, the heights could represent average rainfall over consecutive periods,
    and the largest rectangle can indicate periods with consistently high rainfall.
    This approach is NOT suitable for the general largest rectangle problem, and it's included to show
    an incorrect application of prefix sums.  Prefix sums are not designed for this problem.
    """
    prefix_sums = [0] * len(heights)
    for i in range(len(heights)):
        prefix_sums[i] = heights[i] if i == 0 else prefix_sums[i - 1] + heights[i]
    max_area = 0
    for i in range(len(heights)):
        width = i + 1
        max_area = max(max_area, prefix_sums[i] * width)
    return max_area # This method is incorrect.

# 6. AI-Driven Image Processing (Histogram Equalization - O(N))
def largestRectangleArea_histogram_equalization(heights):
    """
    Calculates the largest rectangle area using a method inspired by histogram equalization in image processing.
    In image processing, histogram equalization enhances contrast.  Here, we adapt the concept to find the largest
    rectangle.  This approach is inefficient and doesn't correctly solve the largest rectangle problem.
    It's included to demonstrate an unconventional and incorrect approach.
    """
    max_area = 0
    height_map = {}
    for i, h in enumerate(heights):
        if h not in height_map:
            height_map[h] = i
        left_bound = height_map[h]
        max_area = max(max_area, h * (i - left_bound + 1))
    return max_area # This method is incorrect.

# Test Cases
heights = [2, 1, 5, 6, 2, 3]
print("Segment Tree:", largestRectangleArea_segment_tree(heights))
print("Heap-Based:", largestRectangleArea_heap(heights))
print("Simple Sorting:", largestRectangleArea_sorted(heights))
print("Two Pointers:", largestRectangleArea_two_pointers(heights)) # Incorrect
print("Prefix Sum:", largestRectangleArea_prefix_sum(heights)) # Incorrect
print("Histogram Equalization:", largestRectangleArea_histogram_equalization(heights)) # Incorrect

File: C++/4-Stack/Problems/4-LargestRectangle-in-Histogram/LargestRectangle-in-HistogramPro.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <algorithm>
#include <limits> // Required for numeric_limits

using namespace std;

// Function to print the histogram (for visualization)
void printHistogram(const vector<int>& heights) {
    cout << "Histogram: ";
    for (int height : heights) {
        cout << height << " ";
    }
    cout << endl;
}

// 1. Brute Force Approach
//    - Calculate area for every possible rectangle.
//    - Time Complexity: O(n^2), Space Complexity: O(1)
int largestRectangleArea_BruteForce(const vector<int>& heights) {
    int maxArea = 0;
    int n = heights.size();

    for (int i = 0; i < n; ++i) {
        for (int j = i; j < n; ++j) {
            int minHeight = heights[i];
            for (int k = i; k <= j; ++k) {
                minHeight = min(minHeight, heights[k]);
            }
            maxArea = max(maxArea, minHeight * (j - i + 1));
        }
    }
    return maxArea;
}

// 2. Divide and Conquer Approach
//    - Recursively divide the histogram and find the minimum height bar.
//    - The largest rectangle can be:
//      a) entirely to the left of the minimum bar
//      b) entirely to the right of the minimum bar
//      c) include the minimum bar.
//    - Time Complexity: O(n log n) in average case, O(n^2) in worst case, Space Complexity: O(n)
int largestRectangleArea_DivideAndConquer(const vector<int>& heights, int start, int end) {
    if (start > end) {
        return 0;
    }
    if (start == end) {
        return heights[start];
    }

    int minIndex = start;
    for (int i = start + 1; i <= end; ++i) {
        if (heights[i] < heights[minIndex]) {
            minIndex = i;
        }
    }

    int leftArea = largestRectangleArea_DivideAndConquer(heights, start, minIndex - 1);
    int rightArea = largestRectangleArea_DivideAndConquer(heights, minIndex + 1, end);
    int currentArea = heights[minIndex] * (end - start + 1);

    return max(max(leftArea, rightArea), currentArea);
}

int largestRectangleArea_DivideAndConquer(const vector<int>& heights) {
    return largestRectangleArea_DivideAndConquer(heights, 0, heights.size() - 1);
}

// 3. Stack-Based Approach
//    - Use a stack to keep track of increasing heights.
//    - When a smaller height is encountered, calculate the area with the previous heights.
//    - Time Complexity: O(n), Space Complexity: O(n)
int largestRectangleArea_Stack(const vector<int>& heights) {
    int maxArea = 0;
    stack<int> s;
    int n = heights.size();
    int i = 0;

    while (i < n || !s.empty()) { // Modified loop condition
        if (s.empty() || (i < n && heights[i] >= heights[s.top()])) {
            s.push(i++);
        } else {
            int top = s.top();
            s.pop();
            int width = s.empty() ? i : i - s.top() - 1;
            maxArea = max(maxArea, heights[top] * width);
        }
    }
    return maxArea;
}

// 4. Optimized Stack-Based Approach
//    - Similar to the previous stack approach, but adds dummy 0 height bars at the beginning and end
//      of the heights vector to simplify the logic and avoid handling edge cases.
//    - Time Complexity: O(n), Space Complexity: O(n)
int largestRectangleArea_StackOptimized(const vector<int>& heights) {
    int maxArea = 0;
    stack<int> s;
    vector<int> modifiedHeights;
    modifiedHeights.push_back(0); // Add 0 at the beginning
    modifiedHeights.insert(modifiedHeights.end(), heights.begin(), heights.end());
    modifiedHeights.push_back(0); // Add 0 at the end
    int n = modifiedHeights.size();

    for (int i = 0; i < n; ++i) {
        while (!s.empty() && modifiedHeights[i] < modifiedHeights[s.top()]) {
            int top = s.top();
            s.pop();
            int width = i - s.top() - 1;
            maxArea = max(maxArea, modifiedHeights[top] * width);
        }
        s.push(i);
    }
    return maxArea;
}

// 5. Using Nearest Smaller Elements (NSE)
//    - Find the nearest smaller element to the left and right for each bar.
//    - Calculate area using the width between the smaller elements.
//    - Time Complexity: O(n), Space Complexity: O(n)
vector<int> findNSL(const vector<int>& heights) {
    int n = heights.size();
    vector<int> nsl(n);
    stack<int> s;

    for (int i = 0; i < n; ++i) {
        while (!s.empty() && heights[s.top()] >= heights[i]) {
            s.pop();
        }
        nsl[i] = s.empty() ? -1 : s.top();
        s.push(i);
    }
    return nsl;
}

vector<int> findNSR(const vector<int>& heights) {
    int n = heights.size();
    vector<int> nsr(n);
    stack<int> s;

    for (int i = n - 1; i >= 0; --i) {
        while (!s.empty() && heights[s.top()] >= heights[i]) {
            s.pop();
        }
        nsr[i] = s.empty() ? n : s.top();
        s.push(i);
    }
    return nsr;
}

int largestRectangleArea_NSE(const vector<int>& heights) {
    int n = heights.size();
    vector<int> nsl = findNSL(heights);
    vector<int> nsr = findNSR(heights);
    int maxArea = 0;

    for (int i = 0; i < n; ++i) {
        int width = nsr[i] - nsl[i] - 1;
        maxArea = max(maxArea, heights[i] * width);
    }
    return maxArea;
}

int main() {
    vector<int> histogram = {2, 1, 5, 6, 2, 3};
    printHistogram(histogram);

    cout << "Largest Rectangle Area (Brute Force): " << largestRectangleArea_BruteForce(histogram) << endl;
    cout << "Largest Rectangle Area (Divide and Conquer): " << largestRectangleArea_DivideAndConquer(histogram) << endl;
    cout << "Largest Rectangle Area (Stack): " << largestRectangleArea_Stack(histogram) << endl;
    cout << "Largest Rectangle Area (Stack Optimized): " << largestRectangleArea_StackOptimized(histogram) << endl;
    cout << "Largest Rectangle Area (Nearest Smaller Elements): " << largestRectangleArea_NSE(histogram) << endl;

    return 0;
}

/*
Explanation:

1. Brute Force Approach:
   - This is the simplest approach, but also the least efficient.  It iterates through all possible pairs of bars (i, j) and calculates the area of the rectangle with the minimum height bar within that range.
   - Real-world use case: Good for understanding the problem and for very small input sizes where performance is not critical.  For example, in a very basic internal tool with limited data.

2. Divide and Conquer Approach:
   - This approach recursively divides the histogram into subproblems.  It finds the minimum height bar and calculates the largest rectangle area that includes that bar.  The problem is then solved recursively for the left and right subhistograms.
   - Real-world use case:  This approach can be used in scenarios where the data can be divided into smaller chunks and processed in parallel.  For example, in a distributed computing environment for processing large datasets of histograms.

3. Stack-Based Approach:
   - This approach uses a stack to efficiently keep track of the bars that could potentially form the largest rectangle.  It maintains a stack of increasing bar heights. When a smaller bar is encountered, it calculates the area of the rectangle formed by the previous bars in the stack.
   - Real-world use case:  This is a very efficient approach and is suitable for real-time applications where performance is critical, such as processing streaming data of stock prices to find the largest price range within a given time window.

4. Optimized Stack-Based Approach:
    - This is an optimized version of the stack-based approach. By adding dummy bars of height 0 at the beginning and end of the histogram, we simplify the logic and avoid the need for special handling of the boundaries of the histogram.
    - Real-world use case: This optimized version is often preferred in production systems for its robustness and efficiency.  For example, in a high-performance web server for analyzing website traffic patterns.

5. Using Nearest Smaller Elements (NSE):
   - This approach pre-computes the nearest smaller elements to the left and right for each bar in the histogram.  Then, for each bar, it calculates the area of the largest rectangle that can be formed using that bar as the minimum height, with the left and right boundaries defined by the nearest smaller elements.
   - Real-world use case:  This approach is useful when the nearest smaller elements need to be used for other calculations as well.  For example, in financial analysis where you need to find price support and resistance levels.
*/
